Optimize `sets.Set`

ash2k opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

Optimize set operations and usage. Please see individual commits.

#### Which issue(s) this PR is related to:
<!--
Please link relevant issues to help with tracking.

To automatically close the linked issue(s) when this PR is merged,
add the word "Fixes" before the issue number or link.
Do not use "Fixes" if the PR is of kind `failing-test` or `flake`.

Reference KEPs when applicable in addition to specific issues.

Examples:
Fixes #<issue number>
<issue link> (issue in a different repository)
KEP: https://github.com/kubernetes/enhancements/issues/<kep-issue-number>

If there is no associated issue, then write "N/A".
-->

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```
/sig api-machinery

ash2k committed:
Use maps.Clone()

ash2k committed:
Use DifferenceSeq() when iterating to avoid materializing the difference as a set

ash2k committed:
Iterate set as-is, no need to turn it into a list

ash2k committed:
Add InsertSet() to allow inserting a set without turning it into a slice

ash2k commented on pkg/proxy/ipvs/ipset.go:
```diff
@@ -164,7 +164,7 @@ func (set *IPSet) syncIPSetEntries() {
 	if !set.activeEntries.Equal(currentIPSetEntries) {
 		// Clean legacy entries
-		for _, entry := range currentIPSetEntries.Difference(set.activeEntries).UnsortedList() {
+		for entry := range currentIPSetEntries.DifferenceIter(set.activeEntries) {
```
The old code was making two allocations (diff set and a slice).

ash2k committed:
Fix NPE

ash2k commented:
This panic looks like a legitimate bug? https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/133231/pull-kubernetes-unit-windows-master/1948977486360481792#1:build-log.txt%3A2478 Fixed in the last commit.

ash2k committed:
Use constructor

ash2k commented:
/retest

adrianmoisey commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -395,9 +395,9 @@ func validateResourceClaimStatusUpdate(status, oldStatus *resource.ResourceClaim
-				if len(newItems) > 0 {
+				for range newSet.DifferenceSeq(oldSet) {
 					allErrs = append(allErrs, field.Forbidden(fldPath.Child("reservedFor"), "new entries may not be added while `deallocationRequested` or `deletionTimestamp` are set"))
+					break // only need to check if there is at lest one item
```
Nit:

```suggestion
					break // only need to check if there is at least one item
```

ash2k replied:
Thanks!

BenTheElder commented:
Do we have benchmark results?

ash2k committed:
Add benchmark for Difference and DifferenceSeq

ash2k commented:
Just pushed a commit with a benchmark. Here is an excerpt with the relevant output:

```
goos: darwin
goarch: arm64
pkg: k8s.io/apimachinery/pkg/util/sets
cpu: Apple M4 Max

BenchmarkStringSet/difference-20
BenchmarkStringSet/difference-20-16                   	  882244	      1335 ns/op	    1636 B/op	       6 allocs/op
BenchmarkStringSet/difference-seq-20
BenchmarkStringSet/difference-seq-20-16               	 2687314	       444.2 ns/op	      56 B/op	       3 allocs/op
BenchmarkStringSet/difference-50
BenchmarkStringSet/difference-50-16                   	  363456	      3267 ns/op	    3457 B/op	       8 allocs/op
BenchmarkStringSet/difference-seq-50
BenchmarkStringSet/difference-seq-50-16               	 1000000	      1164 ns/op	      56 B/op	       3 allocs/op
BenchmarkStringSet/difference-100
BenchmarkStringSet/difference-100-16                  	  190008	      6237 ns/op	    6938 B/op	      10 allocs/op
BenchmarkStringSet/difference-seq-100
BenchmarkStringSet/difference-seq-100-16              	  500174	      2382 ns/op	      56 B/op	       3 allocs/op
BenchmarkStringSet/difference-500
BenchmarkStringSet/difference-500-16                  	   30966	     38250 ns/op	   54276 B/op	      16 allocs/op
BenchmarkStringSet/difference-seq-500
BenchmarkStringSet/difference-seq-500-16              	  118758	      9658 ns/op	      56 B/op	       3 allocs/op
BenchmarkStringSet/difference-1000
BenchmarkStringSet/difference-1000-16                 	   14876	     80534 ns/op	  108803 B/op	      21 allocs/op
BenchmarkStringSet/difference-seq-1000
BenchmarkStringSet/difference-seq-1000-16             	   50556	     23123 ns/op	      56 B/op	       3 allocs/op
```

ash2k commented:
/retest

BenTheElder commented:
thanks -- do we have before/after?

this diff also exposes more APIs, some of which seem to be missing tests.

ash2k commented:
> thanks -- do we have before/after?

I've added the same test to `master` and ran it:

```
goos: darwin
goarch: arm64
pkg: k8s.io/apimachinery/pkg/util/sets
cpu: Apple M4 Max
BenchmarkStringSet
BenchmarkStringSet/difference-20
BenchmarkStringSet/difference-20-16         	  800041	      1353 ns/op	    1636 B/op	       6 allocs/op
BenchmarkStringSet/difference-50
BenchmarkStringSet/difference-50-16         	  354494	      3314 ns/op	    3457 B/op	       8 allocs/op
BenchmarkStringSet/difference-100
BenchmarkStringSet/difference-100-16        	  183486	      6295 ns/op	    6938 B/op	      10 allocs/op
BenchmarkStringSet/difference-500
BenchmarkStringSet/difference-500-16        	   29953	     39356 ns/op	   54291 B/op	      16 allocs/op
BenchmarkStringSet/difference-1000
BenchmarkStringSet/difference-1000-16       	   14665	     81578 ns/op	  108807 B/op	      21 allocs/op
```

> this diff also exposes more APIs, some of which seem to be missing tests.

- I've added a test for pre-existing `Clone()`
- `DifferenceSeq()` is covered by tests for `Difference()` as the former is used by the later.
- I've added a test for `InsetSet()`.

ash2k commented:
To make it easier to see, I've combined the above results below:

```
this PR:

BenchmarkStringSet/difference-20
BenchmarkStringSet/difference-20-16                   	  882244	      1335 ns/op	    1636 B/op	       6 allocs/op
BenchmarkStringSet/difference-50
BenchmarkStringSet/difference-50-16                   	  363456	      3267 ns/op	    3457 B/op	       8 allocs/op
BenchmarkStringSet/difference-100
BenchmarkStringSet/difference-100-16                  	  190008	      6237 ns/op	    6938 B/op	      10 allocs/op
BenchmarkStringSet/difference-500
BenchmarkStringSet/difference-500-16                  	   30966	     38250 ns/op	   54276 B/op	      16 allocs/op
BenchmarkStringSet/difference-1000
BenchmarkStringSet/difference-1000-16                 	   14876	     80534 ns/op	  108803 B/op	      21 allocs/op


master:
BenchmarkStringSet
BenchmarkStringSet/difference-20
BenchmarkStringSet/difference-20-16         	  800041	      1353 ns/op	    1636 B/op	       6 allocs/op
BenchmarkStringSet/difference-50
BenchmarkStringSet/difference-50-16         	  354494	      3314 ns/op	    3457 B/op	       8 allocs/op
BenchmarkStringSet/difference-100
BenchmarkStringSet/difference-100-16        	  183486	      6295 ns/op	    6938 B/op	      10 allocs/op
BenchmarkStringSet/difference-500
BenchmarkStringSet/difference-500-16        	   29953	     39356 ns/op	   54291 B/op	      16 allocs/op
BenchmarkStringSet/difference-1000
BenchmarkStringSet/difference-1000-16       	   14665	     81578 ns/op	  108807 B/op	      21 allocs/op
```

ash2k commented:
/retest

ash2k commented:
/retest

ash2k commented:
/retest

BenTheElder commented:
/cc @thockin

seans3 commented:
/triage accepted

ash2k commented:
I'm going to close this since I don't have the time to debug it. If anyone wants to pick this up and use this code - feel free.
